### PR TITLE
SystemVerilog: Support `break` and `continue`

### PR DESCRIPTION
**Pull request by kbieganski**
The general idea is to do transformations like:

```systemverilog
for (i = 0; i < 100; i++) begin
    <statements>
    continue;
    <statements>
end
```
to
```systemverilog
logic $continue;
for (i = 0; i < 100; i++) begin
    $continue = 0;
    <statements>
    $continue = 1;
    if (!$continue) begin
        <statements>
    end
end
```
and
```systemverilog
for (i = 0; i < 100; i++) begin
    <statements>
    break;
    <statements>
end
```
to
```systemverilog
logic $continue;
logic $break;
$break = 0;
for (i = 0; i < 100; i++) begin
    if (!$break) begin
        $continue = 0;
        <statements>
        $continue = 1;
        if (!$continue) begin
            <statements>
        end
    end
end
```

`while` loops are handled differently in that instead of wrapping the body in an `if`, the break condition gets ANDed with the original loop condition.

Integration PR: https://github.com/antmicro/yosys-uhdm-plugin-integration/pull/866

Signed-off-by: Krzysztof Bieganski <kbieganski@antmicro.com>

### CODE REVIEW

File: systemverilog-plugin/UhdmAst.cc [hunk 1]
```diff
@@ -1151,6 +1151,127 @@ AST::AstNode *UhdmAst::process_value(vpiHandle obj_h)
...
+    }
+    if (break_wire) {
+        auto *break_id = make_ast_node(AST::AST_IDENTIFIER);
+        break_id->str = break_wire->str;
+        // Reset $break before the loop
+        auto *break_assign = make_ast_node(AST::AST_ASSIGN_EQ, {break_id, AST::AstNode::mkconst_int(0, false)});
+        decl_block->children.insert(decl_block->children.begin(), break_assign);
+        decl_block->children.insert(decl_block->children.begin(), break_wire);
+        if (loop->type == AST::AST_REPEAT || loop->type == AST::AST_FOR) {
+            // Wrap loop body in 'if (!break) {...}'
```
Comment (rkapuscik) [inline systemverilog-plugin/UhdmAst.cc:1257]: This works, but I wonder if with `for` shouldn't we amend the loop condition like in `while`?

**Review by rkapuscik — CHANGES_REQUESTED** [review]
Really nice implementation, minor remarks only. Also, I'd say this is big enough to warrant a test or two [for the plugin itself](https://github.com/chipsalliance/yosys-f4pga-plugins/tree/main/systemverilog-plugin/tests), apart from the integration.

File: systemverilog-plugin/UhdmAst.cc (hunk 1 above)
Comment (rkapuscik) [reply systemverilog-plugin/UhdmAst.cc:1257]: Right, conditions in `for` are quite limited. A comment here would be helpful then.

File: systemverilog-plugin/UhdmAst.cc [hunk 2]
```diff
@@ -1151,6 +1151,127 @@ AST::AstNode *UhdmAst::process_value(vpiHandle obj_h)
...
+                if (!continue_wire)
+                    continue_wire = make_cond_var("$continue");
+                auto *continue_id = make_ast_node(AST::AST_IDENTIFIER);
+                continue_id->str = continue_wire->str;
+                block->children.push_back(make_ast_node(AST::AST_ASSIGN_EQ, {continue_id, AST::AstNode::mkconst_int(1, false)}));
+                if (type == AST::AST_BREAK) {
+                    if (!break_wire)
+                        break_wire = make_cond_var("$break");
+                    auto *break_id = make_ast_node(AST::AST_IDENTIFIER);
+                    break_id->str = break_wire->str;
```
[the remark this answers is not in the document]
Comment (kbieganski) [reply systemverilog-plugin/UhdmAst.cc]: Refactored it into a separate function. Only used it in places affected by this PR.

**Comment by kbieganski** [comment]
I added a little test for this. Might not be sufficient, but I'm not sure how comprehensive these plugin tests should be.

**Pull request merged**